marine_benchmark.c

```diff
@@ -0,0 +1,162 @@
+//
+// Created by reznik on 3/29/20.
+//
+#define PACKET_COUNT 210000
+#define CASES 7
+#define PART_LEN PACKET_COUNT / CASES
+
+#include<stdio.h>
+#include<pcap.h>
+#include "marine.h"
+#include <string.h>
+#include <assert.h>
+#include <time.h>
+#include <stdlib.h>
+#include <sys/resource.h>
+#include <zconf.h>
+
+typedef struct {
+    struct pcap_pkthdr *header;
+    const u_char *data;
+} packet;
+
+
+/*
+* Author:  David Robert Nadeau
+* Site:    http://NadeauSoftware.com/
+* License: Creative Commons Attribution 3.0 Unported License
+*          http://creativecommons.org/licenses/by/3.0/deed.en_US
+*/
+size_t get_current_rss(void) {
+    long rss = 0L;
+    FILE *fp = NULL;
+    if ((fp = fopen("/proc/self/statm", "r")) == NULL) {
+        return (size_t) 0L;
+    }
+    if (fscanf(fp, "%*s%ld", &rss) != 1) {
+        fclose(fp);
+        return (size_t) 0L;
+    }
+    fclose(fp);
+    return (size_t) rss * (size_t) sysconf(_SC_PAGESIZE);
+}
+
+int load_cap(char *file, packet packets[]) {
+    printf("Start loading packets from cap\n");
+
+    char errbuff[PCAP_ERRBUF_SIZE];
+    pcap_t *pcap = pcap_open_offline(file, errbuff);
+    if (pcap == NULL) {
+        printf("Error will opening the cap: %s\n", errbuff);
+        return -1;
+    }
+
+    struct pcap_pkthdr *header;
+    const u_char *data;
+    int p_count;
+    for (p_count = 0; p_count < PACKET_COUNT && pcap_next_ex(pcap, &header, &data) >= 0; ++p_count) {
+        packet p = {header, data};
+        packets[p_count] = p;
+    }
+    printf("Cap has been loaded, %d packets were loaded\n", p_count);
+    return p_count;
+}
+
+void benchmark(packet packets[], int part, char *bpf, char *display_filter, char *fields[], int fields_len) {
+    char err_msg[512];
+    int filter_id = marine_add_filter(bpf, display_filter, fields, fields_len, err_msg);
+
+    if (filter_id < 0) {
+        printf("Error creating filter id: %s", err_msg);
+        return;
+    }
+
+    // Splitting the cap into parts so no cache inside wireshark will effect the results
+    int start_index = PART_LEN * part;
```

Comment:
@shaharyarn This is used so every case receives unique packets (in case there are internal caching mechanism which we don't know about and may affect the benchmark)

Reply:
I meant that the packets slicing in order to receive unique packets should be handled outside the benchmark function